components/codec/src/byte.rs

```diff
@@ -341,12 +344,286 @@ impl MemComparableCodecHelper for DescendingMemComparableCodecHelper {
+        self.write_all_bytes(data)
+    }
+}
+
+impl CompactByteEncoder for std::fs::File {
+    #[inline]
+    fn write_compact_bytes(&mut self, data: &[u8]) -> Result<()> {
+        use std::io::Write;
+        let mut buf = [0; super::number::MAX_VARINT64_LENGTH];
+        let written = NumberCodec::encode_var_i64(&mut buf, data.len() as i64);
+        self.write_all(&buf[..written])?;
+        self.write_all(data)?;
+        Ok(())
+    }
+}
+/// Gets the first encoded bytes' length in compactly encoded data.
+///
+/// Compact-encoding includes a VarInt encoded length prefix (1 ~ 9 bytes) and N bytes payload.
+/// This function gets the total bytes length of compact-encoded data, including the length prefix.
+///
+/// Note:
+///     - This function won't check whether the bytes are encoded correctly.
+///     - There can be multiple compact-encoded data, placed one by one. This function only returns
+///       the length of the first one.
+pub fn encoded_compact_len(mut data: &[u8]) -> Result<usize> {
+    let last_encoded = data.as_ptr();
+    let vn = data.read_var_i64()? as usize;
+    unsafe { Ok(vn + data.as_ptr().offset_from(last_encoded) as usize) }
+}
+
+/// Gets the first encoded bytes' length in memory-comparable-encoded data.
+///
+/// Memory-comparable-encoding includes a VarInt encoded length prefix (1 ~ 9 bytes) and N bytes payload.
+/// This function gets the total bytes length of memory-comparable-encoded data, including the length prefix.
+///
+/// Note:
+///     - This function won't check whether the bytes are encoded correctly.
+///     - There can be multiple memory-comparable-encoded data, placed one by one. This function only returns
+///       the length of the first one.
+pub fn encoded_bytes_len(encoded: &[u8], desc: bool) -> usize {
+    let mut idx = MEMCMP_GROUP_SIZE;
+    loop {
+        if encoded.len() < idx + 1 {
+            return encoded.len();
+        }
+        let marker = encoded[idx];
+        if (desc && marker != 0) || (!desc && marker != !0u8) {
+            return idx + 1;
+        }
+        idx += MEMCMP_GROUP_SIZE + 1
+    }
+}
+
+pub trait CompactByteDecoder {
+    /// Decodes bytes which are encoded by `write_compact_bytes` before.
+    fn read_compact_bytes(&mut self) -> Result<Vec<u8>>;
+}
+
+impl<T: NumberDecoder> CompactByteDecoder for T {
+    fn read_compact_bytes(&mut self) -> Result<Vec<u8>> {
+        let vn = self.read_var_i64()? as usize;
+        let data = self.bytes();
+        if data.len() >= vn {
+            let bs = data[0..vn].to_vec();
+            self.advance(vn);
+            return Ok(bs);
+        }
+        Err(Error::UnexpectedEOF)
+    }
+}
+
+impl<T: Read> CompactByteDecoder for std::io::BufReader<T> {
+    fn read_compact_bytes(&mut self) -> Result<Vec<u8>> {
+        let mut buf = [0; super::number::MAX_VARINT64_LENGTH];
+        let mut end = buf.len();
+        for i in 0..super::number::MAX_VARINT64_LENGTH {
+            self.read_exact(&mut buf[i..=i])?;
```

Comment:
could we read exact the first `len=min(super::number::MAX_VARINT64_LENGTH,self.length())` together then iterator to find the stop `u8`?

Reply:
this is implemented for `BufReader` which  don't have length methods except 
1. we read all the bytes. 
2. we implemented it for `std::io::BufReader<File>` , using the file metadata to get the file length